cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -1420,40 +1462,71 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> json_column_to
         cudf::detail::make_device_uvector_async(json_col.string_lengths, stream);
+
+      // Prepare iterator that returns (string_offset, string_length)-tuples
       auto offset_length_it =
         thrust::make_zip_iterator(d_string_offsets.begin(), d_string_lengths.begin());
-      thrust::transform(rmm::exec_policy(stream),
-                        offset_length_it,
-                        offset_length_it + col_size,
-                        d_string_data.data(),
-                        [data = d_input.data()] __device__(auto ip) {
-                          return char_length_pair_t{data + thrust::get<0>(ip), thrust::get<1>(ip)};
-                        });
-      auto str_col_ptr                  = make_strings_column(d_string_data, stream, mr);
-      auto [result_bitmask, null_count] = make_validity(json_col);
-      str_col_ptr->set_null_mask(result_bitmask, null_count);
-      return {std::move(str_col_ptr), {{"offsets"}, {"chars"}}};
+
+      // Prepare iterator that returns (string_offset, string_length)-pairs needed by inference
+      auto string_ranges_it =
+        thrust::make_transform_iterator(offset_length_it, [] __device__(auto ip) {
+          return thrust::pair<json_column::row_offset_t, std::size_t>{
+            thrust::get<0>(ip), static_cast<std::size_t>(thrust::get<1>(ip))};
+        });
+
+      // Prepare iterator that returns (string_ptr, string_length)-pairs needed by type conversion
+      auto string_spans_it = thrust::make_transform_iterator(
+        offset_length_it, [data = d_input.data()] __device__(auto ip) {
+          return thrust::pair<const char*, std::size_t>{
+            data + thrust::get<0>(ip), static_cast<std::size_t>(thrust::get<1>(ip))};
+        });
+
+      // Infer column type
+      auto target_type = cudf::io::detail::infer_data_type(
+        casting_options(options).json_view(), d_input, string_ranges_it, col_size, stream);
+
+      // Convert strings to the inferred data type
+      auto col = cudf::io::json::experimental::detail::parse_data(string_spans_it,
+                                                                  col_size,
+                                                                  target_type,
+                                                                  make_validity(json_col).first,
+                                                                  casting_options(options).view(),
+                                                                  stream,
+                                                                  mr);
+
+      // Reset nullable if we do not have nulls
+      if (col->null_count() == 0) { col->set_null_mask({}); }
+
+      // For string columns return ["offsets", "char"] schema
+      if (target_type.id() == type_id::STRING) {
+        return {std::move(col), {{"offsets"}, {"chars"}}};
+      }
+      // Non-string columns do not have child columns in the schema
```

Comment:
This is a surprising comment without wider context, some further explanation why would be helpful here.

Reply:
Changed to:
> // Non-string leaf-columns (e.g., numeric) do not have child columns in the schema
Hope this is clearer?